File: backend/app/models/card.py
import enum
from datetime import datetime
from sqlalchemy import Column, String, Boolean, DateTime, ForeignKey, UniqueConstraint
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from app.models.base import BaseModel


class CardType(str, enum.Enum):
    CORP = "corp"  # 법인카드
    PERSONAL = "personal"  # 개인카드


class CardIssuer(str, enum.Enum):
    SHINHAN = "shinhan"  # 신한
    KB = "kb"  # KB국민
    HYUNDAI = "hyundai"  # 현대
    SAMSUNG = "samsung"  # 삼성
    LOTTE = "lotte"  # 롯데
    HANA = "hana"  # 하나
    NH = "nh"  # NH농협
    WOORI = "woori"  # 우리
    SC = "sc"  # SC제일
    CITI = "citi"  # 씨티
    OTHER = "other"  # 기타


class Card(BaseModel):
    """결제 카드 관리"""
    __tablename__ = "cards"
    __table_args__ = (
        UniqueConstraint('card_issuer', 'card_number', name='uix_card_unique'),
    )

    card_type = Column(String(20), nullable=False)  # "corp" 또는 "personal"
    card_issuer = Column(String(50), nullable=False)  # 카드사
    card_number = Column(String(20), nullable=False)  # 마지막 4자리 또는 전체 (마스킹)
    owner_name = Column(String(100), nullable=False)  # 카드 소유주명
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True, index=True)
    is_active = Column(Boolean, default=True)
    notes = Column(String(500), nullable=True)  # 비고

    user = relationship("User", foreign_keys=[user_id])

    # 구매 및 판매에서 사용되는 역 관계
    purchases = relationship("Purchase", back_populates="payment_card", foreign_keys="Purchase.payment_card_id")
    sales = relationship("Sale", back_populates="payment_card", foreign_keys="Sale.payment_card_id")

File: frontend/src/types/card.ts
export enum CardType {
  CORP = 'corp',
  PERSONAL = 'personal',
}

export enum CardIssuer {
  SHINHAN = 'shinhan',
  KB = 'kb',
  HYUNDAI = 'hyundai',
  SAMSUNG = 'samsung',
  LOTTE = 'lotte',
  HANA = 'hana',
  NH = 'nh',
  WOORI = 'woori',
  SC = 'sc',
  CITI = 'citi',
  OTHER = 'other',
}

export interface Card {
  id: string;
  card_type: CardType | string;
  card_issuer: CardIssuer | string;
  card_number: string;
  owner_name: string;
  is_active: boolean;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface CardCreate {
  card_type: string;
  card_issuer: string;
  card_number: string;
  owner_name: string;
  notes?: string;
}

export interface CardUpdate {
  card_type?: string;
  card_issuer?: string;
  card_number?: string;
  owner_name?: string;
  is_active?: boolean;
  notes?: string;
}

export const CARD_ISSUER_LABELS: { [key: string]: string } = {
  shinhan: '신한',
  kb: 'KB국민',
  hyundai: '현대',
  samsung: '삼성',
  lotte: '롯데',
  hana: '하나',
  nh: 'NH농협',
  woori: '우리',
  sc: 'SC제일',
  citi: '씨티',
  other: '기타',
};

export const CARD_TYPE_LABELS: { [key: string]: string } = {
  corp: '법인카드',
  personal: '개인카드',
};
